### PR TITLE
feat(skills): edit SKILL frontmatter schema, add CI validation, and documentation

### LINKED ISSUE #623: feat(skills): Update skill frontmatter schema, prompt-builder guidance, CI validation, and documentation
## Overview

Update foundational skill infrastructure to support semantic invocation and optional frontmatter properties before new skills can use the pattern. This is a prerequisite for migrating existing scripts to the Agent Skills pattern.

### 1. Skill Frontmatter Schema Updates

Update `scripts/linting/schemas/skill-frontmatter.schema.json`:

* Add optional properties: `user-invokable` (boolean, default `true`), `disable-model-invocation` (boolean, default `false`), `argument-hint` (string)
* Add `maxLength: 64` to `name`, `maxLength: 1024` to `description`
* Keep `additionalProperties: false`

### 2. Prompt-Builder Instructions Updates

Update `.github/instructions/prompt-builder.instructions.md`:

* Add "Skill Invocation from Callers" subsection under Skill Files
* Document semantic description matching as the primary invocation pattern
* Discourage direct script path references in caller files
* Provide before/after examples showing the transformation from direct paths to semantic task descriptions
* Update Optional Fields to document `user-invokable`, `disable-model-invocation`, `argument-hint` with defaults and behavior

### 3. CI Validation for Skill Structure

Create `scripts/linting/Validate-SkillStructure.ps1`:

* Validate all skills under `.github/skills/` have a `SKILL.md` file
* Validate skill directory structure follows agentskills.io conventions
* Add `npm run validate:skills` script to `package.json`
* Integrate into PR validation workflow as a conditional step — run only when the PR contains changes under `.github/skills/`
* 100% Pester test coverage required — create `scripts/tests/linting/Validate-SkillStructure.Tests.ps1` with full coverage of all validation paths (valid structures, missing SKILL.md, invalid directories, empty skills folder, edge cases)

### 4. Documentation Updates

Update `docs/contributing/skills.md`:

* Add Optional Fields section documenting `argument-hint`, `user-invokable`, `disable-model-invocation`
* Add Invocation Control Matrix table showing setting combinations and their effects
* Add second frontmatter example with optional fields
* Add Semantic Skill Loading section explaining VS Code progressive disclosure (Level 1→2→3)
* Add caller invocation patterns (describe intent, not script paths)
* Update validation checklist with optional field guidance
* Add references to VS Code Agent Skills docs and agentskills.io specification

## References

* [Agent Skills Specification](https://agentskills.io/specification)
* [VS Code Copilot Agent Skills](https://code.visualstudio.com/docs/copilot/customization/agent-skills)
* Existing skill pattern: `.github/skills/video-to-gif/SKILL.md`

## Acceptance Criteria

- [ ] `skill-frontmatter.schema.json` includes optional properties with correct types and constraints
- [ ] `prompt-builder.instructions.md` has "Skill Invocation from Callers" section with examples
- [ ] `Validate-SkillStructure.ps1` validates skill directory structure
- [ ] `Validate-SkillStructure.Tests.ps1` achieves 100% code coverage of the validation script
- [ ] `npm run validate:skills` runs skill validation
- [ ] PR validation workflow runs skill validation only when `.github/skills/` files are changed in the PR
- [ ] `docs/contributing/skills.md` updated with optional fields, invocation control matrix, and semantic loading guidance
- [ ] `npm run lint:frontmatter` passes with existing skills

### PR DESCRIPTION
**Pull request by katriendg**
## Description

Added foundational skill infrastructure to support semantic invocation and optional frontmatter properties. This is a prerequisite for migrating existing scripts to the Agent Skills pattern. Changes span the frontmatter schema, a new CI validation script with Pester tests, prompt-builder guidance for semantic skill invocation, documentation updates, and PR validation pipeline integration.

> Note: Some additional changes were done to `prompt-builder.instructions.md` based on optimization recommended by the `/analyse-prompt` command to improve clarity and structure of the semantic invocation documentation.

- feat(skills): Added `user-invokable`, `disable-model-invocation`, and `argument-hint` optional properties to `skill-frontmatter.schema.json` with `maxLength` constraints on `name` (64), `description` (1024), and `argument-hint` (256)
- feat(skills): Created `Validate-SkillStructure.ps1` (464 lines) with SKILL.md presence checks, frontmatter field validation, name-directory consistency, unrecognized subdirectory warnings, changed-files-only mode via Git merge-base, CI annotations, and JSON results export
- feat(skills): Created `Validate-SkillStructure.Tests.ps1` (876 lines) with full Pester coverage across `Get-SkillFrontmatter`, `Test-SkillDirectory`, `Get-ChangedSkillDirectories`, and `Write-SkillValidationResults` including edge cases for git failures, path normalization, and CI annotation emission
- feat(skills): Added "Skill Invocation from Callers" subsection to `prompt-builder.instructions.md` documenting semantic description matching as the primary invocation pattern with before/after examples, progressive disclosure levels, and updated optional field documentation
- docs(skills): Updated `docs/contributing/skills.md` with Optional Fields section, Invocation Control Matrix table, frontmatter example with optional fields, Semantic Skill Loading section with progressive disclosure levels, caller invocation patterns, and external references to agentskills.io and VS Code docs
- ci(skills): Created `skill-validation.yml` reusable workflow with `soft-fail` and `changed-files-only` inputs, pinned action SHAs, and artifact upload for `skill-validation-results`
- ci(skills): Integrated `skill-validation` job into `pr-validation.yml` with `changed-files-only: true`
- chore: Added `validate:skills` npm script to `package.json` and appended it to `lint:all`; updated `CONTRIBUTING.md`, `scripts/README.md`, `scripts/linting/README.md`, `docs/architecture/workflows.md`, `docs/contributing/ai-artifacts-common.md`, `.github/workflows/README.md`, and `.github/PULL_REQUEST_TEMPLATE.md` to document the new validation step
- chore: Added `agentskills` and `invokable` to `.cspell.json` dictionary

## Related Issue(s)

Closes #623

## Type of Change

Select all that apply:

**Code & Documentation:**

- [ ] Bug fix (non-breaking change fixing an issue)
- [x] New feature (non-breaking change adding functionality)
- [ ] Breaking change (fix or feature causing existing functionality to change)
- [x] Documentation update

**Infrastructure & Configuration:**

- [x] GitHub Actions workflow
- [ ] Linting configuration (markdown, PowerShell, etc.)
- [ ] Security configuration
- [ ] DevContainer configuration
- [ ] Dependency update

**AI Artifacts:**

- [x] Reviewed contribution with `prompt-builder` agent and addressed all feedback
- [x] Copilot instructions (`.github/instructions/*.instructions.md`)
- [ ] Copilot prompt (`.github/prompts/*.prompt.md`)
- [ ] Copilot agent (`.github/agents/*.agent.md`)
- [ ] Copilot skill (`.github/skills/*/SKILL.md`)

> **Note for AI Artifact Contributors**:
>
> - **Agents**: Research, indexing/referencing other project (using standard VS Code GitHub Copilot/MCP tools), planning, and general implementation agents likely already exist. Review `.github/agents/` before creating new ones.
> - **Skills**: Must include both bash and PowerShell scripts. See [Skills](../docs/contributing/skills.md).
> - **Model Versions**: Only contributions targeting the **latest Anthropic and OpenAI models** will be accepted. Older model versions (e.g., GPT-3.5, Claude 3) will be rejected.
> - See [Agents Not Accepted](../docs/contributing/custom-agents.md#agents-not-accepted) and [Model Version Requirements](../docs/contributing/ai-artifacts-common.md#model-version-requirements).

**Other:**

- [x] Script/automation (`.ps1`, `.sh`, `.py`)
- [ ] Other (please describe):

## Testing

- Ran `/analyze-prompt`
- Ran `Validate-SkillStructure.Tests.ps1` with Pester to verify all tests pass, including edge cases for git failures, path normalization, and CI annotation emission. 
- Ran `npm run validate:skills` to ensure the validation script executes without errors in a real-world scenario (based on currently one existing skill).

### Required Checks

- [x] Documentation is updated (if applicable)
- [x] Files follow existing naming conventions
- [ ] Changes are backwards compatible (if applicable)
- [x] Tests added for new functionality (if applicable)

### AI Artifact Contributions

- [x] Used `/prompt-analyze` to review contribution
- [x] Addressed all feedback from `prompt-builder` review
- [x] Verified contribution follows common standards and type-specific requirements

### Required Automated Checks

The following validation commands must pass before merging:

- [x] Markdown linting: `npm run lint:md`
- [x] Spell checking: `npm run spell-check`
- [x] Frontmatter validation: `npm run lint:frontmatter`
- [x] Skill structure validation: `npm run validate:skills`
- [x] Link validation: `npm run lint:md-links`
- [x] PowerShell analysis: `npm run lint:ps`

## Security Considerations

- [x] This PR does not contain any sensitive or NDA information
- [ ] Any new dependencies have been reviewed for security issues
- [ ] Security-related scripts follow the principle of least privilege

## Additional Notes

All four deliverables from issue #623 are addressed in this PR:

1. **Skill Frontmatter Schema Updates** — Added three optional properties with type constraints and length limits
2. **Prompt-Builder Instructions Updates** — Added semantic invocation guidance with before/after examples and progressive disclosure documentation
3. **CI Validation for Skill Structure** — Created `Validate-SkillStructure.ps1` with comprehensive Pester test coverage in `Validate-SkillStructure.Tests.ps1`
4. **Documentation Updates** — Updated `docs/contributing/skills.md` with optional fields, invocation control matrix, semantic skill loading, and external specification references

🛠️ - Generated by Copilot

### CODE REVIEW

File: scripts/tests/linting/Validate-SkillStructure.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,876 @@
...
+        }
+
+        return Get-Item $skillDir
+    }
+}
+
+AfterAll {
+    if ($script:TempTestDir -and (Test-Path $script:TempTestDir)) {
+        Remove-Item -Path $script:TempTestDir -Recurse -Force -ErrorAction SilentlyContinue
+    }
```
[the remark this answers is not in the document]
Comment (katriendg) [reply scripts/tests/linting/Validate-SkillStructure.Tests.ps1:67]: done

File: scripts/linting/Validate-SkillStructure.ps1 [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+    }
+
+    # Check scripts/ subdirectory contents
+    $scriptsDirPath = Join-Path -Path $Directory.FullName -ChildPath 'scripts'
+    if (Test-Path $scriptsDirPath -PathType Container) {
+        $scriptFiles = Get-ChildItem -Path $scriptsDirPath -File -ErrorAction SilentlyContinue |
+            Where-Object { $_.Extension -in @('.ps1', '.sh') }
+        if ($null -eq $scriptFiles -or @($scriptFiles).Count -eq 0) {
+            $warnings.Add("'scripts/' subdirectory exists but contains no .ps1 or .sh files in '$relativePath'")
+        }
```
[the remark this answers is not in the document]
Comment (katriendg) [reply scripts/linting/Validate-SkillStructure.ps1:200]: fixed

File: docs/contributing/skills.md [hunk 3]
```diff
@@ -305,8 +388,9 @@ Run these commands before submission:
 
 ```bash
 npm run lint:frontmatter      # Validate SKILL.md frontmatter
+npm run validate:skills       # Validate skill directory structure
 npm run psscriptanalyzer      # Validate PowerShell scripts
```
[the remark this answers is not in the document]
Comment (katriendg) [reply docs/contributing/skills.md]: fixed

File: scripts/linting/Validate-SkillStructure.ps1 [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+        if (-not $hasPowerShell -and -not $hasBash) {
+            $errors.Add("'scripts/' subdirectory exists but contains no .ps1 or .sh files in '$relativePath'")
+        }
+        elseif (-not $hasPowerShell) {
+            $errors.Add("'scripts/' subdirectory is missing a required .ps1 file in '$relativePath'")
+        }
+        elseif (-not $hasBash) {
+            $errors.Add("'scripts/' subdirectory is missing a required .sh file in '$relativePath'")
+        }
+    }
```
[the remark this answers is not in the document]
Comment (katriendg) [reply scripts/linting/Validate-SkillStructure.ps1:201]: did updates to required and optional in code and all documentation

File: .github/workflows/skill-validation.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+          persist-credentials: false
+          fetch-depth: 0
+
+      - name: Run skill validation
+        shell: pwsh
+        run: |
+          $params = @{}
+          if ('${{ inputs.changed-files-only }}' -eq 'true') {
+            $params['ChangedFilesOnly'] = $true
+            $params['BaseBranch'] = 'origin/${{ github.base_ref }}'
```
[the remark this answers is not in the document]
Comment (katriendg) [reply .github/workflows/skill-validation.yml]: updated to follow default main as per other workflows and script callers

File: .github/instructions/prompt-builder.instructions.md [hunk 6]
```diff
@@ -352,7 +404,7 @@ Optional fields available by file type:
 
 * `tools:` - Tool restrictions for agents and subagents. When omitted, all tools are accessible. When specified, list only tools available in the current VS Code context.
 * `handoffs:` - Agent handoff declarations. Each entry includes `label` (display text, supports emoji), `agent` (target agent name), and optionally `prompt` (slash command to invoke) and `send` (boolean, auto-send the prompt when `true`).
-* `user-invocable:` - Boolean. Set to `false` to hide the agent from the user and prevent direct invocation. Defaults to `true` when omitted. Use for subagents that should not appear in the agent picker.
+* `user-invocable:` - Boolean. Set to `false` to hide the artifact from the user and prevent direct invocation. Defaults to `true` when omitted. Use for subagents that should not appear in the agent picker or background-only skills that should not appear in the slash command menu.
```
[the remark this answers is not in the document]
Comment (katriendg) [reply .github/instructions/prompt-builder.instructions.md:407]: VS code only supports user-invocable going forward, updating schema and docs
